### PR TITLE
[PLAT-8185] Limit retries of event payloads based on size and age, and session payloads based on age

### PR DESCRIPTION
**Pull request by kstenerud**
## Goal

To prevent the buildup of old or large files, event and session files will be deleted on a failed send if they meet certain criteria:

- Events if they are older than 60 days or larger than 1MB
- Sessions if they are older than 60 days

## Testing

 New e2e tests

### CODE REVIEW

File: bugsnag-android-core/src/main/java/com/bugsnag/android/SessionFilenameInfo.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            )
+        }
+
+        private fun findUuidInFilename(file: File): String {
+            return file.name.substring(0, 35)
+        }
+
+        fun findTimestampInFilename(file: File): Long {
+            try {
+                return java.lang.Long.parseLong(file.name.substring(file.name.indexOf("_")))
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/SessionFilenameInfo.kt:49]: I might be missing something, but it doesn't look like this took the UUID into account?

```suggestion
            return file.name.substring(36, file.name.indexOf("_")).toLongOrNull() ?: 0
```

File: features/fixtures/mazerunner/jvm-scenarios/src/main/java/com/bugsnag/android/mazerunner/scenarios/DiscardOldEventsScenario.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    init {
+        config.launchDurationMillis = 0
+    }
+
+    fun setEventFileTimestamp(file: File, timestamp: Long) {
+        val name = file.name
+        val suffix = name.substring(name.indexOf("_"))
+        val dstFile = File(file.parent, "" + timestamp + suffix)
+        if (!file.renameTo(dstFile)) {
+            throw RuntimeException("Could not rename " + file.canonicalPath + " to " + dstFile.canonicalPath)
```
Comment (lemnik) [inline features/fixtures/mazerunner/jvm-scenarios/src/main/java/com/bugsnag/android/mazerunner/scenarios/DiscardOldEventsScenario.kt:25]: nit:

```suggestion
            throw RuntimeException("Could not rename ${file.canonicalPath} to ${dstFile.canonicalPath}")
```

File: features/fixtures/mazerunner/jvm-scenarios/src/main/java/com/bugsnag/android/mazerunner/scenarios/DiscardOldEventsScenario.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+) : Scenario(config, context, eventMetadata) {
+
+    init {
+        config.launchDurationMillis = 0
+    }
+
+    fun setEventFileTimestamp(file: File, timestamp: Long) {
+        val name = file.name
+        val suffix = name.substring(name.indexOf("_"))
+        val dstFile = File(file.parent, "" + timestamp + suffix)
```
Comment (lemnik) [inline features/fixtures/mazerunner/jvm-scenarios/src/main/java/com/bugsnag/android/mazerunner/scenarios/DiscardOldEventsScenario.kt:23]: nit:

```suggestion
        val suffix = name.substringAfter("_")
        val dstFile = File(file.parent, "${timestamp}_$suffix")
```

**Comment by twometresteve** [comment]
As this touches some of the low-level routines, I would ensure that a full CI run has been performed before merging this PR (just include `[full ci]` in the commit message).

**Comment by twometresteve** [comment]
@kstenerud I hope you don't mind, but I have taken the liberty of updating updating your branch from `next`, as I want to chain a reformatting changing onto this PR (all should become apparent shortly!).

**Review by lemnik — APPROVED** [review]
LGTM

**Pull request merged**